### PR TITLE
Feature/test training

### PR DESCRIPTION
**Pull request by luismiguel0529**
## Summary
- Test Book Controller
- Test User Controller
- Test User Dao 
- Test Book Dao
## Trello Card
https://trello.com/c/mNbyy49Y/28-junit-testing

### CODE REVIEW

File: src/test/java/wolox/training/repositories/BookRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        List<Book> bookList = bookRepository.findAll();
+        assertEquals(bookList.get(0).getAuthor(), manyTestBooks.get(0).getAuthor());
+        assertEquals(bookList.get(0).getId(), manyTestBooks.get(0).getId());
+    }
+
+    @Test
+    void whenUpdateBookThenReturnBookUpdated() {
+        oneTestBook.setAuthor("miguel");
+        bookRepository.save(oneTestBook);
+        assertEquals(oneTestBook.getAuthor(), oneTestBook.getAuthor());
```
Comment (facucompe) [inline src/test/java/wolox/training/repositories/BookRepositoryTest.java]: This is redundant. Use a findById to check the book is updated

File: src/test/java/wolox/training/repositories/UserRepositoryTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        List<User> userList = usersRepository.findAll();
+        assertEquals(userList.get(0).getName(), manyTestUsers.get(0).getName());
+        assertEquals(userList.get(0).getId(), manyTestUsers.get(0).getId());
+    }
+
+    @Test
+    void whenUpdateUserThenReturnUserUpdated() {
+        oneTestUser.setName("miguel");
+        User persistedUser = usersRepository.save(oneTestUser);
+        assertEquals(oneTestUser.getName(), persistedUser.getName());
```
Comment (facucompe) [inline src/test/java/wolox/training/repositories/UserRepositoryTest.java]: Same here

File: src/test/java/wolox/training/util/TestEntities.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+
+public class TestEntities {
+
+    public static User mockOneUser() {
+        User oneTestUser = new User();
+        oneTestUser.setId(1l);
```
Comment (facucompe) [inline src/test/java/wolox/training/util/TestEntities.java]: Don't set the Id

**Review by facucompe — CHANGES_REQUESTED** [review]

File: src/main/java/wolox/training/controllers/UsersController.java [hunk 4]
```diff
@@ -94,6 +94,7 @@ public User findById(@PathVariable Long id) {
     @PostMapping
     @ResponseStatus(HttpStatus.CREATED)
     public User create(@RequestBody User user) {
+
```
Comment (joredudiaz) [inline src/main/java/wolox/training/controllers/UsersController.java]: Can remove this break line please?

File: src/test/java/wolox/training/controllers/BookControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                .contentType(MediaType.APPLICATION_JSON)
+                .characterEncoding("utf-8")
+                .content(json))
+                .andDo(print())
+                .andExpect(status().isCreated());
+    }
+
+    @Test
+    @DisplayName("Test, When a book is updated , it return status OK")
+    void whenUpdateBookThenReturnStatusCreated() throws Exception {
```
Comment (joredudiaz) [inline src/test/java/wolox/training/controllers/BookControllerTest.java:91]: Can you test please when update a book that does not exists and returns not found?

File: src/test/java/wolox/training/controllers/BookControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                .contentType(MediaType.APPLICATION_JSON)
+                .characterEncoding("utf-8")
+                .content(json))
+                .andDo(print())
+                .andExpect(status().isOk());
+    }
+
+    @Test
+    @DisplayName("Test, When a book is deleted , it return status No Content")
+    void whenDeleteBookThenReturnStatusNoContent() throws Exception {
```
Comment (joredudiaz) [inline src/test/java/wolox/training/controllers/BookControllerTest.java:119]: Can you test please when delete a book that does not exists and returns not found?

File: src/test/java/wolox/training/controllers/UserControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+@RunWith(SpringRunner.class)
+@WebMvcTest(controllers = UsersController.class)
+class UserControllerTest {
+
+    @Autowired
+    private MockMvc mvc;
+
+    @MockBean
+    private UsersRepository mockUsersRepository;
```
Comment (joredudiaz) [inline src/test/java/wolox/training/controllers/UserControllerTest.java:38]: Can you add a break line please?

File: src/test/java/wolox/training/controllers/UserControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    @Test
+    @DisplayName("Test find all user ,return status OK")
+    void whenFindUserByIdThenReturnStatusOK() throws Exception {
+        given(mockUsersRepository.findById(1L)).willReturn(Optional.of(oneTestUser));
+        String url = (USER_PATH + "/1");
+        mvc.perform(get(url)
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.name").value(oneTestUser.getName()));
+
```
Comment (joredudiaz) [inline src/test/java/wolox/training/controllers/UserControllerTest.java]: Can you remove this break line please?

File: src/test/java/wolox/training/controllers/UserControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+                .contentType(MediaType.APPLICATION_JSON)
+                .characterEncoding("utf-8")
+                .content(json))
+                .andDo(print())
+                .andExpect(status().isCreated());
+    }
+
+    @Test
+    @DisplayName("Test, When a user is updated , it return status OK")
+    void whenUpdateUserThenReturnStatusCreated() throws Exception {
```
Comment (joredudiaz) [inline src/test/java/wolox/training/controllers/UserControllerTest.java:107]: Can you test please when update an user that does not exists and returns not found?

File: src/test/java/wolox/training/controllers/UserControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        given(mockUsersRepository.findById(1L)).willReturn(Optional.empty());
+        String url = (USER_PATH + "/1");
+        mvc.perform(get(url)
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isNotFound());
+    }
+
+    @Test
+    @DisplayName("Test,When a users is searched ,it return status OK")
+    void whenFindAllUserTHenReturnStatusOK() throws Exception {
```
Comment (joredudiaz) [inline src/test/java/wolox/training/controllers/UserControllerTest.java]: `whenFindAllUserThenReturnStatusOK`

File: src/test/java/wolox/training/controllers/UserControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+                .contentType(MediaType.APPLICATION_JSON)
+                .characterEncoding("utf-8")
+                .content(json))
+                .andDo(print())
+                .andExpect(status().isOk());
+    }
+
+    @Test
+    @DisplayName("Test, When a user is deleted , it return status No Content")
+    void whenDeleteUserThenReturnStatusNoContent() throws Exception {
```
Comment (joredudiaz) [inline src/test/java/wolox/training/controllers/UserControllerTest.java:135]: Can you test please when delete an user that does not exists and returns not found?

File: src/test/java/wolox/training/controllers/BookControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        given(mockBookRepository.findById(1L)).willReturn(Optional.empty());
+        String url = (USER_PATH + "/1");
+        mvc.perform(get(url)
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isNotFound());
+    }
+
+    @Test
+    @DisplayName("Test,When a books is searched ,it return status OK")
+    void whenFindAllBookTHenReturnStatusOK() throws Exception {
```
Comment (joredudiaz) [inline src/test/java/wolox/training/controllers/BookControllerTest.java]: `whenFindAllBookThenReturnStatusOK`

File: src/test/java/wolox/training/repositories/BookRepositoryTest.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+    @BeforeAll
+    static void setUp() {
+        oneTestBook = TestEntities.mockBook();
+        manyTestBooks = TestEntities.mockManyBooks();
+        twoTestBook = TestEntities.mockBookToPersis();
+    }
+
+    @Test
+    void whenCallFindALlBookThenReturnAllList() {
```
Comment (joredudiaz) [inline src/test/java/wolox/training/repositories/BookRepositoryTest.java]: `whenCallFindAllBookThenReturnAllList`

File: src/test/java/wolox/training/repositories/UserRepositoryTest.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+    @BeforeAll
+    static void setUp() {
+        oneTestUser = TestEntities.mockOneUser();
+        manyTestUsers = TestEntities.mockManyUsers();
+        twoTestUser = TestEntities.mockUserToPersis();
+    }
+
+    @Test
+    void whenCallFindALlUserThenReturnAllList() {
```
Comment (joredudiaz) [inline src/test/java/wolox/training/repositories/UserRepositoryTest.java]: `whenCallFindAllUserThenReturnAllList`

**Review by joredudiaz — COMMENTED** [review]
😄 a couple comments

**Review by facucompe — APPROVED** [review]

**Review by facucompe — APPROVED** [review]

**Review by joredudiaz — APPROVED** [review]

**Pull request merged**